### PR TITLE
Abacus: version and feature update

### PR DESCRIPTION
**Pull request by yizeyi18**
Package abacus had been created before, but is long outdated. This PR updates new versions of the package, and implements build and run features as in https://github.com/deepmodeling/abacus-develop/issues/3291.
@spackbot fix style

### CODE REVIEW

**Comment by yizeyi18** [comment]
@spackbot fix style

**Comment by yizeyi18** [comment]
@spackbot fix style

**Comment by yizeyi18** [comment]
@spackbot fix style
Why local `spack style --fix` always failing...

**Review by caic99 — APPROVED** [review]

**Review by bernhardkaindl — COMMENTED** [review]
@yizeyi18: When I try to build, I get:
```py
bin/spack install -v abacus
==> Error: Package 'spack.pkg.builtin.nvpl-blas' not found.
Use 'spack create' to create a new package.

Did you mean one of the following packages?
  py-liblas
  netlib-xblas
  liblas
```
This seems to be common error I also get from another update - how to build?

**Comment by yizeyi18** [comment]
@bernhardkaindl I'm also confused by that...Actually, after updating my fork and before editing abacus the package, this error jumps out. Maybe any bug in CudaPackages class?

For a temporary fix, try `spack clean -m` then install again. That works on my pc.

**Review by bernhardkaindl — DISMISSED** [review]
@yizeyi18 Thanks for checking and mentioning that you can build it after `spack clean -m` on your laptop!

I don't have a CUDA compiler installed, and my next try ended in an error that it could not find my CUDA compiler.

I then went on to try `bin/spack install -v abacus~cuda`, which resulted in:
```py
     35    -- Looking for a CUDA compiler
     36    -- Looking for a CUDA compiler - NOTFOUND
  >> 37    CMake Error at /home/t_bernhardk/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/cmake-3.27.9-qdzb3zo5s3c6v
           f4oj7p4jevv4qdda4cy/share/cmake-3.27/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
     38      Could NOT find IntelMKL (missing: MKL_INTEL MKL_INTEL_THREAD MKL_CORE
     39      MKL_SCALAPACK MKL_BLACS_INTELMPI MKL_INCLUDE_DIR)
     40    Call Stack (most recent call first):
     41      /home/t_bernhardk/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/cmake-3.27.9-qdzb3zo5s3c6vf4oj7p4jevv4q
           dda4cy/share/cmake-3.27/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
     42      cmake/FindIntelMKL.cmake:24 (find_package_handle_standard_args)
     43      CMakeLists.txt:385 (find_package)
```
I guess I'd have to install `intel-mkl` externally to build it, or select it to be downloaded and installed by `spack`?

I am just trying to understand the process with the aim of finding out how CI could test the build of packages like abacus correctly, thanks!

**Comment by yizeyi18** [comment]
Hi @bernhardkaindl,
`abacus` do not require mkl compulsively. Which `blas` provider do you use?
According to packages.py and CMakeLists.txt, only when you use blas among `intel-mkl`, `intel-parallel-studio`, `intel-oneapi-mkl` would spack define `MKLROOT` in command line and cmake would try to find a `mkl`; otherwise cmake would try find a `LAPACK`. May you show your `spack-build-out.txt`, `spack-build-env-mods.txt` and `spack-build-env.txt`? 

EDIT: One spec sample of `abacus`, concreted on my wsl:
<details><summary>Sample spec</summary>
<p>

```shell
$ spack spec -l abacus
Input spec
--------------------------------
 -   abacus

Concretized
--------------------------------
 -   nbwv4cw  abacus@3.5.4%gcc@13.2.0~cuda+elpa~ipo+lcao+libxc~mathlib+mpi+openmp~rocm~tests build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-skylake
[+]  uhgc4yl      ^cereal@1.3.2%gcc@13.2.0~ipo build_system=cmake build_type=Release generator=make patches=2dfa0bf arch=linux-ubuntu22.04-skylake
[e]  5jtet3y      ^cmake@3.22.1%gcc@13.2.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-ubuntu22.04-skylake
[+]  pw5jd2b      ^elpa@2023.05.001%gcc@13.2.0~autotune~cuda+mpi+openmp~rocm build_system=autotools arch=linux-ubuntu22.04-skylake
[+]  sjr3v5h          ^libtool@2.4.7%gcc@13.2.0 build_system=autotools arch=linux-ubuntu22.04-skylake
[+]  tdejqfr              ^m4@1.4.19%gcc@13.2.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-ubuntu22.04-skylake
[e]  nodstrc                  ^diffutils@3.8%gcc@13.2.0 build_system=autotools arch=linux-ubuntu22.04-skylake
[+]  wq3ghrz                  ^libsigsegv@2.14%gcc@13.2.0 build_system=autotools arch=linux-ubuntu22.04-skylake
[e]  zg3y6s6          ^python@3.10.6%gcc@13.2.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=0d98e93,7d40923,f2fd060 arch=linux-ubuntu22.04-skylake
[+]  yvepezs      ^fftw@3.3.10%gcc@13.2.0+mpi+openmp~pfft_patches build_system=autotools precision=double,float arch=linux-ubuntu22.04-skylake
[+]  eonvvfe      ^gcc-runtime@13.2.0%gcc@13.2.0 build_system=generic arch=linux-ubuntu22.04-skylake
[e]  6cnijfi      ^gmake@4.3%gcc@13.2.0~guile build_system=autotools patches=599f134 arch=linux-ubuntu22.04-skylake
[+]  xqbit2c      ^libxc@6.2.2%gcc@13.2.0~cuda~kxc~lxc+shared build_system=autotools arch=linux-ubuntu22.04-skylake
[+]  4kvryat          ^autoconf@2.69%gcc@13.2.0 build_system=autotools patches=35c4492,7793209,a49dd5b arch=linux-ubuntu22.04-skylake
[+]  344ajhj          ^automake@1.16.5%gcc@13.2.0 build_system=autotools arch=linux-ubuntu22.04-skylake
[+]  x6tczzn          ^perl@5.38.0%gcc@13.2.0+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=linux-ubuntu22.04-skylake
[+]  bbwnq6p              ^berkeley-db@18.1.40%gcc@13.2.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu22.04-skylake
[+]  aynbov5              ^bzip2@1.0.8%gcc@13.2.0~debug~pic+shared build_system=generic arch=linux-ubuntu22.04-skylake
[+]  2cxlzy6              ^gdbm@1.23%gcc@13.2.0 build_system=autotools arch=linux-ubuntu22.04-skylake
[+]  epwc5ug                  ^readline@8.2%gcc@13.2.0 build_system=autotools patches=bbf97f1 arch=linux-ubuntu22.04-skylake
[+]  5szvra6      ^netlib-scalapack@2.2.0%gcc@13.2.0~ipo~pic+shared build_system=cmake build_type=Release generator=make patches=072b006,1c9ce5f,244a9aa arch=linux-ubuntu22.04-skylake
[+]  wgxa4ex      ^openblas@0.3.23%gcc@13.2.0~bignuma~consistent_fpcsr+fortran~ilp64+locking+pic+shared build_system=makefile symbol_suffix=none threads=openmp arch=linux-ubuntu22.04-skylake
[+]  5nthj2e      ^openmpi@4.1.5%gcc@13.2.0~atomics~cuda~cxx~cxx_exceptions~gpfs~internal-hwloc~internal-pmix~java~legacylaunchers~lustre~memchecker~openshmem~orterunprefix+romio+rsh~singularity+static+vt+wrapper-rpath build_system=autotools fabrics=none schedulers=none arch=linux-ubuntu22.04-skylake
[+]  sg627sl          ^hwloc@2.9.1%gcc@13.2.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml~oneapi-level-zero~opencl+pci~rocm build_system=autotools libs=shared,static arch=linux-ubuntu22.04-skylake
[+]  f2n3i6i              ^libpciaccess@0.17%gcc@13.2.0 build_system=autotools arch=linux-ubuntu22.04-skylake
[+]  scbnhrk                  ^util-macros@1.19.3%gcc@13.2.0 build_system=autotools arch=linux-ubuntu22.04-skylake
[+]  rftdk7o              ^libxml2@2.10.3%gcc@13.2.0~python build_system=autotools arch=linux-ubuntu22.04-skylake
[+]  7gfjnwe                  ^libiconv@1.17%gcc@13.2.0 build_system=autotools libs=shared,static arch=linux-ubuntu22.04-skylake
[+]  hri3lgq                  ^xz@5.4.1%gcc@13.2.0~pic build_system=autotools libs=shared,static arch=linux-ubuntu22.04-skylake
[+]  ghcbbga              ^ncurses@6.4%gcc@13.2.0~symlinks+termlib abi=none build_system=autotools arch=linux-ubuntu22.04-skylake
[+]  43w6aco          ^numactl@2.0.14%gcc@13.2.0 build_system=autotools patches=4e1d78c,62fc8a8,ff37630 arch=linux-ubuntu22.04-skylake
[e]  6nqbdun          ^openssh@8.9p1%gcc@13.2.0+gssapi build_system=autotools arch=linux-ubuntu22.04-skylake
[+]  xm23fkm          ^pkgconf@1.9.5%gcc@13.2.0 build_system=autotools arch=linux-ubuntu22.04-skylake
[+]  oaj4xlz          ^pmix@4.2.4%gcc@13.2.0~docs+pmi_backwards_compatibility~python~restful build_system=autotools arch=linux-ubuntu22.04-skylake
[+]  snzahvl              ^libevent@2.1.12%gcc@13.2.0+openssl build_system=autotools arch=linux-ubuntu22.04-skylake
[+]  ernmtd7                  ^openssl@3.1.2%gcc@13.2.0~docs+shared build_system=generic certs=mozilla arch=linux-ubuntu22.04-skylake
[+]  2zfwgve                      ^ca-certificates-mozilla@2023-05-30%gcc@13.2.0 build_system=generic arch=linux-ubuntu22.04-skylake
[+]  saszwn4          ^zlib-ng@2.1.3%gcc@13.2.0+compat+opt build_system=autotools patches=299b958,ae9077a,b692621 arch=linux-ubuntu22.04-skylake
```

</p>
</details>

File: var/spack/repos/builtin/packages/abacus/package.py [hunk 1]
```diff
@@ -6,40 +6,214 @@
...
+    version("3.1.3", sha256="ca056628472c8959f93f78e8e4fe10e1ea51a0b3333cef0c8c841dd22f9c0d1b")
+    version("3.1.2", sha256="0fcebc659754deecfa912391a14ee9749c84ec45849e8b231ad9c765602826d4")
+    version("3.1.1", sha256="a268a28e8dce2da98260265cb8edd0a16dd4a307a5016022c5cd2598e9df93c9")
+    version("3.1.0", sha256="f038e20c712276c5a7c8f26ed6af6ff5d7c57110accd45882d11b1d64dee3f90")
+    version("3.0.5", sha256="81ff1b8f10909eeda2d888f97f615e5b92aa26c101a32f36f346f78aa7638ba9")
+    version("3.0.4", sha256="b01714e734b0aa5b0b52f3d7871e4bc9f6c81f25a444b1287e3f2e98a4f5d289")
+    version("3.0.3", sha256="a4ae83cdb2c5029f152064f163d1b66b265980793d76b79908c073027401d2fc")
+    version("3.0.2", sha256="b0c3aad9cac18d0ca73c78c69326031407af6a4bfcc953b3a27506c298bc59a3")
+    version("3.0.1", sha256="812941146c31ab53c9a7695abbca6bfb36ae55878e38f8d0f13a17f3d9c36dc0")
+    version("3.0.0", sha256="1c1299f53788beb2f6b6180d8484d584b6f293c81cea25f45b9632dd608ba4f9")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/abacus/package.py]: Do we need all these versions? Or can we get by with the latest patch for each minor?

File: var/spack/repos/builtin/packages/abacus/package.py [hunk 2]
```diff
@@ -6,40 +6,214 @@
...
+    variant(
+        "elpa",
+        default=True,
+        description="Enable optimised diagonalisation routines from ELPA",
+        when="+lcao",
+    )
+    variant("mathlib", default=False, description="Enable ABACUS's builtin libm")
+    variant(
+        "tests", default=False, description="Build ABACUS unit tests", when="build_system=cmake"
+    )
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/abacus/package.py]: Can we avoid the variant and use `self.run_tests` instead? Or is this variant installing something in the prefix?

File: var/spack/repos/builtin/packages/abacus/package.py [hunk 3]
```diff
@@ -6,40 +6,214 @@
...
+    variant(
+        "benchmarks",
+        default=False,
+        description="Enable ABACUS's builtin benchmark tests",
+        when="+tests",
+    )
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/abacus/package.py]: Same question here.

File: var/spack/repos/builtin/packages/abacus/package.py [hunk 4]
```diff
@@ -6,40 +6,214 @@
...
+    variant("rocm", default=False, description="(Experimental)Enable rocm support")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/abacus/package.py]: This one should be inherited from the base class, so there is no need to add the variant again:
```suggestion
```

File: var/spack/repos/builtin/packages/abacus/package.py [hunk 5]
```diff
@@ -6,40 +6,214 @@
...
+        depends_on("libxc")
+    with when("+lcao"):
+        depends_on("cereal")
+        depends_on("scalapack")
+    with when("+elpa"):
+        depends_on("elpa")
+    with when("+tests"):
+        depends_on("googletest")
+    with when("+benchmarks"):
+        depends_on("benchmark")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/abacus/package.py]: I guess these are `type="test"`, even though we have a few long-standing issues with modeling test deps. :thinking:

File: var/spack/repos/builtin/packages/abacus/package.py [hunk 6]
```diff
@@ -6,40 +6,214 @@
...
+    with when("+openmp"):
+        depends_on("fftw+openmp", when="^[virtuals=fftw-api] fftw")
+        depends_on("elpa+openmp", when="+elpa")
+        depends_on(
+            "openblas threads=openmp",
+            when="^[virtuals=blas] openblas" or "^[virtuals=lapack] openblas",
+        )
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/abacus/package.py]: This would evaluate to the first arg, so you can simply use:
```suggestion
        depends_on(
            "openblas threads=openmp",
            when="^[virtuals=blas] openblas",
        )
```
As far as I know `openblas` provides lapack and blas together (it cannot build on top of other blas libraries).

File: var/spack/repos/builtin/packages/abacus/package.py [hunk 7]
```diff
@@ -6,40 +6,214 @@
...
+    with when("~openmp"):
+        depends_on("elpa~openmp", when="+elpa")
+
+    requires("%clang" or "%rocmcc", when="+rocm", msg="build with rocm requires rocm compiler")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/abacus/package.py]: This is equivalent to:
```suggestion
    requires("%clang", when="+rocm", msg="build with rocm requires rocm compiler")
```
due to how `or` works with strings in Python

File: var/spack/repos/builtin/packages/abacus/package.py [hunk 8]
```diff
@@ -6,40 +6,214 @@
...
-    depends_on("elpa+openmp", when="+openmp")
-    depends_on("elpa~openmp", when="~openmp")
-    depends_on("cereal")
-    depends_on("libxc")
-    depends_on("fftw")
-    # MPI is a necessary dependency
-    depends_on("mpi", type=("build", "link", "run"))
-    depends_on("mkl")
 
+class MakefileBuilder(makefile.MakefileBuilder):
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/abacus/package.py:152]: The signature of the method below needs to be adjusted:
```
def edit(self, pkg, spec, prefix):
    ...
```
etc.

File: var/spack/repos/builtin/packages/abacus/package.py [hunk 9]
```diff
@@ -98,3 +272,62 @@ def edit(self, spec, prefix):
 
     def install(self, spec, prefix):
         install_tree("bin", prefix.bin)
+
+
+class CMakeBuilder(cmake.CMakeBuilder):
+    def cmake_args(self):
+        spec = self.spec
+        args = []
+        args += [
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/abacus/package.py]:
```suggestion
        args = [
```

**Review by alalazo — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/abacus/package.py (hunk 1 above)
Comment (yizeyi18) [reply var/spack/repos/builtin/packages/abacus/package.py]: I followed `llvm`'s version-including strategy; patching minors is also acceptable. As a fast-releasing package, I suppose including versions directly.

File: var/spack/repos/builtin/packages/abacus/package.py (hunk 2 above)
Comment (yizeyi18) [reply var/spack/repos/builtin/packages/abacus/package.py]: Nothing installed. What is self.run_tests?

File: var/spack/repos/builtin/packages/abacus/package.py (hunk 4 above)
Comment (yizeyi18) [reply var/spack/repos/builtin/packages/abacus/package.py]: This feature is not ready for production use; pointing this out in `spack info abacus` requires writing down the variant explicitly.

File: var/spack/repos/builtin/packages/abacus/package.py (hunk 5 above)
Comment (yizeyi18) [reply var/spack/repos/builtin/packages/abacus/package.py]: Yes. Spack has a dependent type `test` in addition to `build, link, run`?

File: var/spack/repos/builtin/packages/abacus/package.py (hunk 6 above)
Comment (yizeyi18) [reply var/spack/repos/builtin/packages/abacus/package.py]: In some corner cases, openblas could be concretized as only a lapack provider, e. g.: 
<details><summary>one corner case example</summary>
<p>

```
yizeyi18@DESKTOP-5UK5DD4:~/soft/spack$ spack spec -l abacus ^netlib-xblas
Input spec
--------------------------------
 -   abacus
 -       ^netlib-xblas

Concretized
--------------------------------
 -   q44hmr5  abacus@develop%gcc@13.2.0~cuda+elpa~ipo+lcao+libxc~mathlib+mpi+openmp~rocm~tests build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-skylake
[+]  uhgc4yl      ^cereal@1.3.2%gcc@13.2.0~ipo build_system=cmake build_type=Release generator=make patches=2dfa0bf arch=linux-ubuntu22.04-skylake
[e]  5jtet3y      ^cmake@3.22.1%gcc@13.2.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-ubuntu22.04-skylake
 -   xccstl6      ^elpa@2023.05.001%gcc@13.2.0~autotune~cuda+mpi+openmp~rocm build_system=autotools arch=linux-ubuntu22.04-skylake
[+]  sjr3v5h          ^libtool@2.4.7%gcc@13.2.0 build_system=autotools arch=linux-ubuntu22.04-skylake
[+]  tdejqfr              ^m4@1.4.19%gcc@13.2.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-ubuntu22.04-skylake
[e]  nodstrc                  ^diffutils@3.8%gcc@13.2.0 build_system=autotools arch=linux-ubuntu22.04-skylake
[+]  wq3ghrz                  ^libsigsegv@2.14%gcc@13.2.0 build_system=autotools arch=linux-ubuntu22.04-skylake
[e]  zg3y6s6          ^python@3.10.6%gcc@13.2.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=0d98e93,7d40923,f2fd060 arch=linux-ubuntu22.04-skylake
[+]  eonvvfe      ^gcc-runtime@13.2.0%gcc@13.2.0 build_system=generic arch=linux-ubuntu22.04-skylake
[e]  6cnijfi      ^gmake@4.3%gcc@13.2.0~guile build_system=autotools patches=599f134 arch=linux-ubuntu22.04-skylake
[e]  c4hwk6l      ^intel-oneapi-mkl@2023.2.0%gcc@13.2.0~cluster+envmods~ilp64+shared build_system=generic mpi_family=none threads=none arch=linux-ubuntu22.04-skylake
[e]  vkj5jcu      ^intel-oneapi-mpi@2021.10.0%gcc@13.2.0+envmods~external-libfabric~generic-names~ilp64 build_system=generic arch=linux-ubuntu22.04-skylake
[+]  xqbit2c      ^libxc@6.2.2%gcc@13.2.0~cuda~kxc~lxc+shared build_system=autotools arch=linux-ubuntu22.04-skylake
[+]  4kvryat          ^autoconf@2.69%gcc@13.2.0 build_system=autotools patches=35c4492,7793209,a49dd5b arch=linux-ubuntu22.04-skylake
[+]  344ajhj          ^automake@1.16.5%gcc@13.2.0 build_system=autotools arch=linux-ubuntu22.04-skylake
[+]  x6tczzn          ^perl@5.38.0%gcc@13.2.0+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=linux-ubuntu22.04-skylake
[+]  bbwnq6p              ^berkeley-db@18.1.40%gcc@13.2.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu22.04-skylake
[+]  aynbov5              ^bzip2@1.0.8%gcc@13.2.0~debug~pic+shared build_system=generic arch=linux-ubuntu22.04-skylake
[+]  2cxlzy6              ^gdbm@1.23%gcc@13.2.0 build_system=autotools arch=linux-ubuntu22.04-skylake
[+]  epwc5ug                  ^readline@8.2%gcc@13.2.0 build_system=autotools patches=bbf97f1 arch=linux-ubuntu22.04-skylake
[+]  ghcbbga                      ^ncurses@6.4%gcc@13.2.0~symlinks+termlib abi=none build_system=autotools arch=linux-ubuntu22.04-skylake
[+]  xm23fkm                          ^pkgconf@1.9.5%gcc@13.2.0 build_system=autotools arch=linux-ubuntu22.04-skylake
[+]  saszwn4              ^zlib-ng@2.1.3%gcc@13.2.0+compat+opt build_system=autotools patches=299b958,ae9077a,b692621 arch=linux-ubuntu22.04-skylake
 -   cyx2ffx      ^netlib-scalapack@2.2.0%gcc@13.2.0~ipo~pic+shared build_system=cmake build_type=Release generator=make patches=072b006,1c9ce5f,244a9aa arch=linux-ubuntu22.04-skylake
 -   omuynna      ^netlib-xblas@1.0.248%gcc@13.2.0+fortran+plain_blas build_system=autotools arch=linux-ubuntu22.04-skylake
[+]  wgxa4ex      ^openblas@0.3.23%gcc@13.2.0~bignuma~consistent_fpcsr+fortran~ilp64+locking+pic+shared build_system=makefile symbol_suffix=none threads=openmp arch=linux-ubuntu22.04-skylake
```

</p>
</details> 

In this example, `netlib-xblas`, `openblas`, `netlib-scalapack` and `intel-oneapi-mkl` serves as provider of `blas`, `lapack`, `scalapack` and `fftw-api`.

File: var/spack/repos/builtin/packages/abacus/package.py (hunk 8 above)
Comment (yizeyi18) [reply var/spack/repos/builtin/packages/abacus/package.py:152]: resolved by 484229d.

File: var/spack/repos/builtin/packages/abacus/package.py (hunk 5 above)
Comment (yizeyi18) [reply var/spack/repos/builtin/packages/abacus/package.py]: solved by 3c96d82.

File: var/spack/repos/builtin/packages/abacus/package.py (hunk 4 above)
Comment (alalazo) [reply var/spack/repos/builtin/packages/abacus/package.py]: Fair.

**Review by caic99 — APPROVED** [review]

File: var/spack/repos/builtin/packages/abacus/package.py [hunk 10]
```diff
@@ -6,43 +6,218 @@
...
+    version(
+        "2.3.4",
+        sha256="16b817311c8055cf055021d895642fbb913585f08ec96f9aff3e0ef20148ec0e",
+        deprecated=True,
+    )
+    version(
+        "2.3.3",
+        sha256="a1246c4a6de385826cae15d9a446e862331eb189a592622b1014c7b49504c908",
+        deprecated=True,
+    )
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/abacus/package.py]: Let's not add versions that start as deprecated :wink:

File: var/spack/repos/builtin/packages/abacus/package.py (hunk 2 above)
Comment (alalazo) [reply var/spack/repos/builtin/packages/abacus/package.py]: It's a variable that is set to `True` when the package is installed asking for build-time tests. You can grep the builtin repo to see how it's used. One example would be `hdf5`, for instance.

File: var/spack/repos/builtin/packages/abacus/package.py (hunk 6 above)
Comment (alalazo) [reply var/spack/repos/builtin/packages/abacus/package.py]: What I mean is:
```
>>> "foo" or "bar"
'foo'
```
so the argument above evaluates as the first operand

File: var/spack/repos/builtin/packages/abacus/package.py (hunk 10 above)
Comment (yizeyi18) [reply var/spack/repos/builtin/packages/abacus/package.py]: OK. Removed by 45ce442fc.

File: var/spack/repos/builtin/packages/abacus/package.py (hunk 1 above)
Comment (caic99) [reply var/spack/repos/builtin/packages/abacus/package.py]: Hi @yizeyi18 ,
Actually ABACUS does not strictly follows semantic versioning, which means that "patch versions" also contains new features and modifications that might breaks forward compatibility.
I think it would be better if you can keep all those patch versions.

File: var/spack/repos/builtin/packages/abacus/package.py (hunk 2 above)
Comment (yizeyi18) [reply var/spack/repos/builtin/packages/abacus/package.py]: According to private discussion with ABACUS developers, unit tests are temporarily removed in this PR.

File: var/spack/repos/builtin/packages/abacus/package.py (hunk 6 above)
Comment (yizeyi18) [reply var/spack/repos/builtin/packages/abacus/package.py]: removed in [f38aa6b](https://github.com/spack/spack/pull/42974/commits/f38aa6b1ff91d5cae79b06ca34314a3e4ee2687b) .

File: var/spack/repos/builtin/packages/abacus/package.py (hunk 1 above)
Comment (yizeyi18) [reply var/spack/repos/builtin/packages/abacus/package.py]: Outdated patch versions removed.

**Review by caic99 — APPROVED** [review]

**Comment by yizeyi18** [comment]
Hi @alalazo! The issues you raised have been resolved. Could you please re-review the updated pull request? Thanks!

File: var/spack/repos/builtin/packages/abacus/package.py [hunk 11]
```diff
@@ -101,3 +208,60 @@ def edit(self, spec, prefix):
...
+
+
+class CMakeBuilder(cmake.CMakeBuilder):
+    def cmake_args(self):
+        spec = self.spec
+        args = [
+            self.define_from_variant("ENABLE_MPI", "mpi"),
+            self.define_from_variant("USE_OPENMP", "openmp"),
+            self.define_from_variant("USE_ELPA", "elpa"),
+            self.define_from_variant("USE_ABACUS_LIBM", "mathlib"),
```
Comment (bendamasio) [inline var/spack/repos/builtin/packages/abacus/package.py:220]:
```py
==> abacus: Executing phase: 'cmake'
==> Error: KeyError: '"mathlib" is not a variant of "abacus"'
```
Either remove setting `USE_ABACUS_LIBM` from mathlib, set it to a fixed value or define the variant for it.

It sounds like to set if an abacus-internal libm shall be used.

Option 1, to set a fixed value, would possibly be like:
```suggestion
    self.define("USE_ABACUS_LIBM", False),
```
I guess if the system does have a "good" (performant?) libm / libc, then you may not need to use an internal libm.

But it's your expert decision what to do with it. You can also leave it to auto-detect.
If that is just the standard `libm` installed on the systems,
and you do not need to turn it off or on, your may just remove it and let abacus  decide.

BTW, are all these new variants supported by all abacus versions?

**Review by bendamasio — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/abacus/package.py (hunk 11 above)
Comment (bendamasio) [reply var/spack/repos/builtin/packages/abacus/package.py:220]: @yizeyi18:
1. I think it should be common sense to build what you push and request review for.
2. In the interest of saving face on this PR, please check that your pushes do build and work.

As that care has not yet been taken by you, I'm setting the "draft" status, to prevent spending GitLab CI compute resources on this PR until you fix the build and all requested changes are resolved.

When the PR is fixed, describe which tests you have done (including on which platforms you built it) before requesting reviews again.
Thanks a lot!

**Review by bernhardkaindl — CHANGES_REQUESTED** [review]
@caic99 As you requested my review: Please delay requesting a review from me unless / until you can show that the updated package works.

**Comment by alalazo** [comment]
@yizeyi18 Are you still interested in this PR?

**Comment by yizeyi18** [comment]
@alalazo Not having much time. My tutor forbids me to spend much time on open source projects like abacus/spack😢

**Comment by yizeyi18** [comment]
As spack packages are moved to another repository, this PR (at spack/spack) is closed.
